### ISSUE TITLE
High CPU overhead when using ParallelExecutor in a minimal throttled example

### ISSUE THREAD

**Issue by maun** [bevyengine/bevy#4718]
## Bevy version

0.8-dev main aced6aff

## Operating system & version

macOS 12.3.1

## What you did

Used minimal example:

```
fn main() {
    App::new()
        .add_plugins(MinimalPlugins)
        .add_system(sleep_system)
        .run();
}

fn sleep_system() {
    std::thread::sleep(std::time::Duration::from_millis(15));
}
```
and measured CPU usage with `top` / Activity Monitor.
The app was started using `cargo run --release`.

## What you expected to happen

Running with close to 0% CPU usage.

## What actually happened

CPU usage of 10-15%

## Additional information

Using `SingleThreadedExecutor` by hardcoding it in [bevy_ecs](https://github.com/bevyengine/bevy/blob/aced6aff041555b97d97c96381b551527569881f/crates/bevy_ecs/src/schedule/stage.rs#L124)
```
pub fn parallel() -> Self {
    // Self::new(Box::new(ParallelExecutor::default()))
    Self::new(Box::new(SingleThreadedExecutor::default()))
}
```
results in a CPU usage of about 0.3%.

Similar behaviour exists for an empty window without using the renderer feature and also in the breakout example. For the empty window the CPU usage was quite low, whereas in the breakout example the CPU usage dropped about >10% with the change, but was still unexpectedly high.

I recorded [flame graphs](https://github.com/flamegraph-rs/flamegraph), to get the interactive version download the svg and open them in a browser.

### ParallelExecutor
![flamegraph-multi](https://user-images.githubusercontent.com/40381850/167709059-516140de-7bd9-415f-bad4-378fa7d09a47.svg)

The call to `std::thread::sleep` is at the highest "tower" on the left, you can zoom in when using a local interactive svg.

### SingleThreadedExecutor
![flamegraph-single](https://user-images.githubusercontent.com/40381850/167709143-2d72ed06-ae2d-47be-b24c-1ef2e9068827.svg)

The same test without the sleep system results in 150% CPU usage with the `ParallelExecutor` and close to 100% with the `SingleThreadedExecutor `.

Issue #1462 is probably related but was closed and is less specific.

**Comment by alice-i-cecile** [comment]
This is an excellent issue report. I'd like to try and benchmark this; we have a long way to go here.

**Comment by james7132** [comment]
@maun can you please try again with the changes from #4740? This rewrite a large portion of the internals of our TaskPool, which seems to have a lot of its internals showing up in your flamegraphs. In particular, the way that scope is handled (with the local executor's try_tick usage) should be reworked to use proper suspend and `Waker` infrastructure.

**Comment by maun** [comment]
With the changes from #4740 I get a CPU usage of about 2.3%. This is the flamegraph for the PR:
![flamegraph_pr](https://user-images.githubusercontent.com/40381850/170653142-2e6d6555-05cb-45c6-b564-bdde36ff861f.svg)

**Comment by james7132** [comment]
> With the changes from #4740 I get a CPU usage of about 2.3%. This is the flamegraph for the PR: ![flamegraph_pr](https://user-images.githubusercontent.com/40381850/170653142-2e6d6555-05cb-45c6-b564-bdde36ff861f.svg)

This might be an issue. We're currently investigating some overhead with that PR, which may be very prominently shown in your flamegraph as parking the thread, whereas main is currently spin-waiting on the results. Is a reduction from 15% idle to 2.3% acceptable here for us to close the issue or push this further? At some point this becomes a tradeoff question of latency vs resource usage.

**Comment by DigitalJokerMan** [comment]
+1 on this issue. I'd like to use Bevy for a command-line game server and running a barebones App with just MinimalPlugins results in using as many CPU resources as the machine has left.

**Comment by ypoonse** [comment]
> +1 on this issue. I'd like to use Bevy for a command-line game server and running a barebones App with just MinimalPlugins results in using as many CPU resources as the machine has left.

Same here, very similar same scenario: a headless server that runs game logic. In my case I only have a system that needs to run once every 0.1s, so my setup looks like this:

```rust
    App::new()
        .add_plugins(MinimalPlugins)
        .add_system_set(
            SystemSet::new()
                .with_run_criteria(FixedTimestep::step(GAME_TIMESTEP))
                .with_system(logic_system1),
        )
```

But in this case, it's also still consuming a full cpu.

One possible workaround is to take control of the scheduling myself:

```rust
let mut world = World::default();

let mut update_stage = SystemStage::parallel();
update_stage.add_system(my_system);

let entity1 = world
    .spawn()
    .insert(...)
    .id();

// Adding resources
world.insert_resource(....);

// Run systems every 100 ms, I haven't tested this part
loop {
    update_stage.run(&mut world);
    std::thread::sleep(std::time::Duration::from_millis(100));
}
```

**Comment by hymm** [comment]
@ypoonse for your usage, I would consider using the bevy_framepace crate or write a custom runner. https://github.com/bevyengine/bevy/blob/v0.8.0/examples/app/custom_loop.rs

**Comment by martinfiers** [comment]
Thanks you very much @hymm ! CPU just went down with a lot by using a custom loop.

**Comment by aqez** [comment]
An alternative, what worked for me after scouring the code, is to specify a specific wait time in the ScheduleRunnerSettings:
```rust

use bevy::{
    app::{RunMode, ScheduleRunnerSettings}
};

.insert_resource(ScheduleRunnerSettings {
    run_mode: RunMode::Loop {
        wait: Some(Duration::from_millis(20)),
    },
})

```

**Comment by alice-i-cecile** [comment]
I'm going to close this out as resolved then :)

**Comment by maun** [comment]
> I'm going to close this out as resolved then :)

I thought the issue is still present, for example in the breakout example? I just used the MinimalPlugins for a minimal example, not because this is only relevant for game servers.

**Comment by james7132** [comment]
A fix in #4718 might work as it shifts the spin-ticking behavior of `TaskPool::scope` to use async_executor's `Executor::run` which will park the main thread while it's sitting idle.
